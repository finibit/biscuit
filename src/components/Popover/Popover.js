import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import PopoverArrow from './parts/PopoverArrow'
import PopoverContent from './parts/PopoverContent'

const PopoverStyled = styled.div`
	box-sizing: border-box;
	position: relative;
	display: inline-block;
`

const Popover = (props) => {
	const {
		trigger,
		content,
		width,
		height,
		padding,
		...rest
	} = props

	return (
		<PopoverStyled
			{...rest}
		>
			{trigger}
			<PopoverContent
				$element="Popover"
				$parent={PopoverStyled}
				$color={0}
				$bgColor={1}
				$margin="none"
				$padding={padding}
				$width={width}
				$height={height}
				$border={0}
				$elevation={2}
			>
				{content}
				<PopoverArrow
					$element="Popover"
					$bgColor={1}
					$elevation={1}
				/>
			</PopoverContent>
		</PopoverStyled>
	)
}

Popover.propTypes = {
	/** Popover trigger. */
	trigger: PropTypes.element.isRequired,

	/** Popover content. */
	content: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.element]),

	/** Fixed popover width. */
	width: PropTypes.string,

	/** Fixed popover height. */
	height: PropTypes.string,

	/** The amount of padding around the popover. */
	padding: PropTypes.oneOfType([
		PropTypes.number,
		PropTypes.string,
		PropTypes.object,
		PropTypes.array,
	]),
}

Popover.defaultProps = {
	content: null,
	width: 'initial',
	height: 'initial',
	padding: 2,
}

export default Popover
