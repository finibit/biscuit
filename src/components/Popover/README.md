### Usage

#### Placement

```jsx
<Popover trigger={<span>Hover me</span>} content="Hello!" placement="top-left" />
```

```jsx
<Popover trigger={<span>Hover me</span>} content="Hello!" placement="top" />
```

```jsx
<Popover trigger={<span>Hover me</span>} content="Hello!" placement="top-right" />
```

```jsx
<Popover trigger={<span>Hover me</span>} content="Hello!" placement="right-top" />
```

```jsx
<Popover trigger={<span>Hover me</span>} content="Hello!" placement="right" />
```

```jsx
<Popover trigger={<span>Hover me</span>} content="Hello!" placement="right-bottom" />
```

```jsx
<Popover trigger={<span>Hover me</span>} content="Hello!" placement="bottom-right" />
```

```jsx
<Popover trigger={<span>Hover me</span>} content="Hello!" placement="bottom" />
```

```jsx
<Popover trigger={<span>Hover me</span>} content="Hello!" placement="bottom-left" />
```

```jsx
<Popover trigger={<span>Hover me</span>} content="Hello!" placement="left-bottom" />
```

```jsx
<Popover trigger={<span>Hover me</span>} content="Hello!" placement="left" />
```

```jsx
<Popover trigger={<span>Hover me</span>} content="Hello!" placement="left-top" />
```

#### Visible

```jsx
<Popover trigger={<span>Always visible</span>} content="Hello!" placement="right-top" visible />
```
