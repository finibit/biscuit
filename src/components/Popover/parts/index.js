export { default as PopoverArrow } from './PopoverArrow'
export { default as PopoverContent } from './PopoverContent'
