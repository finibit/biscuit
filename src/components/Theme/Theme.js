import React from 'react'
import PropTypes from 'prop-types'
import { ThemeProvider, createGlobalStyle } from 'styled-components'
import { mergeTheme, baseTheme, themeGet } from '../../themes'
import types from '../../types'

const GlobalStyle = createGlobalStyle`
	html, body {
		width: 100%;
		height: 100%;
		margin: 0;
		padding: 0;
	}

	body {
		color: ${themeGet.color('black', 'inherit')};
		font-family: ${themeGet.fontFamily('secondary')};
		-webkit-font-smoothing: antialiased;
	}
`

const Theme = (props) => {
	const {
		theme,
		global,
	} = props

	return (
		<ThemeProvider theme={mergeTheme(baseTheme, theme)}>
			<div>
				{ global && <GlobalStyle /> }
				{props.children}
			</div>
		</ThemeProvider>
	)
}

Theme.propTypes = {
	global: PropTypes.bool,
	...types.theme,
}

Theme.defaultProps = {
	global: true,
	theme: baseTheme,
}

export default Theme
