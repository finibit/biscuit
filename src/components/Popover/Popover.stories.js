import React, { useState } from 'react'
import { storiesOf } from '@storybook/react'
import Wrapper from '../../Wrapper'
import Popover from './Popover'
import Text from '../Text'

const content = "Intelligent beings dream of the mind's eye tendrils of gossamer clouds circumnavigated billions upon billions a mote of dust suspended in a sunbeam?"

const PopoverStateful = (props) => {
  const [isVisible, setIsVisible] = useState(false)

  return (
    <Popover
      isVisible={isVisible}
      onShow={() => setIsVisible(true)}
      onHide={() => setIsVisible(false)}
      {...props}
    />
  )
}

storiesOf('Popover', module)
  .addDecorator(Wrapper)
  .add('with placement', () => (
    <>
      <PopoverStateful
        placement='top-left'
        trigger={<Text mr={3}>Top left</Text>}
        content={content}
      />
      <PopoverStateful
        placement='top'
        trigger={<Text mr={3}>Top</Text>}
        content={content}
      />
      <PopoverStateful
        placement='top-right'
        trigger={<Text mr={3}>Top right</Text>}
        content={content}
      />
      <PopoverStateful
        placement='right-top'
        trigger={<Text mr={3}>Right top</Text>}
        content={content}
      />
      <PopoverStateful
        placement='right'
        trigger={<Text mr={3}>Right</Text>}
        content={content}
      />
      <PopoverStateful
        placement='right-bottom'
        trigger={<Text mr={3}>Right bottom</Text>}
        content={content}
      />
      <PopoverStateful
        placement='bottom-right'
        trigger={<Text mr={3}>Bottom right</Text>}
        content={content}
      />
      <PopoverStateful
        placement='bottom'
        trigger={<Text mr={3}>Bottom</Text>}
        content={content}
      />
      <PopoverStateful
        placement='bottom-left'
        trigger={<Text mr={3}>Bottom left</Text>}
        content={content}
      />
      <PopoverStateful
        placement='left-bottom'
        trigger={<Text mr={3}>Left bottom</Text>}
        content={content}
      />
      <PopoverStateful
        placement='left'
        trigger={<Text mr={3}>Left</Text>}
        content={content}
      />
      <PopoverStateful
        placement='left-top'
        trigger={<Text mr={3}>Left top</Text>}
        content={content}
      />
    </>
  ))
