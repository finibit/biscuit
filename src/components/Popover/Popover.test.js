import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Theme from '../Theme'
import Popover from './Popover'

describe('Popover', () => {
  test('with placement', () => {

  })
})