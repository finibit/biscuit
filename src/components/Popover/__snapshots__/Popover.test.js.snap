// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover renders content 1`] = `
.c3 {
  background-color: #ffffff;
  width: 12px;
  height: 12px;
  position: absolute;
  left: 18px;
  bottom: 0;
  -webkit-transform: translateY(50%) rotate(45deg);
  -ms-transform: translateY(50%) rotate(45deg);
  transform: translateY(50%) rotate(45deg);
  box-shadow: 1px 1px 0 0 rgba(0,0,0,0.2);
}

.c2 {
  margin: 0;
  padding: 1rem;
  color: #000000;
  background-color: #ffffff;
  border: 1px solid #d8d8d8;
  box-shadow: 0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);
  display: none;
  position: absolute;
  z-index: 99;
  width: initial;
  height: initial;
  left: 0;
  bottom: 0;
  -webkit-transform: translateY(-56px);
  -ms-transform: translateY(-56px);
  transform: translateY(-56px);
}

.c0:hover .c2 {
  display: block;
}

.c1 {
  box-sizing: border-box;
  position: relative;
  display: inline-block;
}

<div>
  <div
    className="c0 c1"
  >
    <h1>
      Trigger
    </h1>
    <div
      className="c2"
    >
      <p>
        Content
      </p>
      <div
        className="c3"
      />
    </div>
  </div>
</div>
`;

exports[`Popover renders trigger 1`] = `
.c3 {
  background-color: #ffffff;
  width: 12px;
  height: 12px;
  position: absolute;
  left: 18px;
  bottom: 0;
  -webkit-transform: translateY(50%) rotate(45deg);
  -ms-transform: translateY(50%) rotate(45deg);
  transform: translateY(50%) rotate(45deg);
  box-shadow: 1px 1px 0 0 rgba(0,0,0,0.2);
}

.c2 {
  margin: 0;
  padding: 1rem;
  color: #000000;
  background-color: #ffffff;
  border: 1px solid #d8d8d8;
  box-shadow: 0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);
  display: none;
  position: absolute;
  z-index: 99;
  width: initial;
  height: initial;
  left: 0;
  bottom: 0;
  -webkit-transform: translateY(-56px);
  -ms-transform: translateY(-56px);
  transform: translateY(-56px);
}

.c0:hover .c2 {
  display: block;
}

.c1 {
  box-sizing: border-box;
  position: relative;
  display: inline-block;
}

<div>
  <div
    className="c0 c1"
  >
    <h1>
      Trigger
    </h1>
    <div
      className="c2"
    >
      <div
        className="c3"
      />
    </div>
  </div>
</div>
`;

exports[`Popover renders with custom padding 1`] = `
.c3 {
  background-color: #ffffff;
  width: 12px;
  height: 12px;
  position: absolute;
  left: 18px;
  bottom: 0;
  -webkit-transform: translateY(50%) rotate(45deg);
  -ms-transform: translateY(50%) rotate(45deg);
  transform: translateY(50%) rotate(45deg);
  box-shadow: 1px 1px 0 0 rgba(0,0,0,0.2);
}

.c0:hover .c4 {
  display: block;
}

.c0:hover .c5 {
  display: block;
}

.c0:hover .c6 {
  display: block;
}

.c2 {
  margin: 0;
  padding: 10rem;
  color: #000000;
  background-color: #ffffff;
  border: 1px solid #d8d8d8;
  box-shadow: 0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);
  display: none;
  position: absolute;
  z-index: 99;
  width: initial;
  height: initial;
  left: 0;
  bottom: 0;
  -webkit-transform: translateY(-56px);
  -ms-transform: translateY(-56px);
  transform: translateY(-56px);
}

.c0:hover .c2 {
  display: block;
}

.c1 {
  box-sizing: border-box;
  position: relative;
  display: inline-block;
}

<div>
  <div
    className="c0 c1"
  >
    <h1>
      Trigger
    </h1>
    <div
      className="c2"
    >
      <p>
        Content
      </p>
      <div
        className="c3"
      />
    </div>
  </div>
</div>
`;

exports[`Popover renders with fixed height 1`] = `
.c3 {
  background-color: #ffffff;
  width: 12px;
  height: 12px;
  position: absolute;
  left: 18px;
  bottom: 0;
  -webkit-transform: translateY(50%) rotate(45deg);
  -ms-transform: translateY(50%) rotate(45deg);
  transform: translateY(50%) rotate(45deg);
  box-shadow: 1px 1px 0 0 rgba(0,0,0,0.2);
}

.c0:hover .c4 {
  display: block;
}

.c0:hover .c5 {
  display: block;
}

.c2 {
  margin: 0;
  padding: 1rem;
  color: #000000;
  background-color: #ffffff;
  border: 1px solid #d8d8d8;
  box-shadow: 0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);
  display: none;
  position: absolute;
  z-index: 99;
  width: initial;
  height: 100px;
  left: 0;
  bottom: 0;
  -webkit-transform: translateY(-56px);
  -ms-transform: translateY(-56px);
  transform: translateY(-56px);
}

.c0:hover .c2 {
  display: block;
}

.c1 {
  box-sizing: border-box;
  position: relative;
  display: inline-block;
}

<div>
  <div
    className="c0 c1"
  >
    <h1>
      Trigger
    </h1>
    <div
      className="c2"
    >
      <p>
        Content
      </p>
      <div
        className="c3"
      />
    </div>
  </div>
</div>
`;

exports[`Popover renders with fixed width 1`] = `
.c3 {
  background-color: #ffffff;
  width: 12px;
  height: 12px;
  position: absolute;
  left: 18px;
  bottom: 0;
  -webkit-transform: translateY(50%) rotate(45deg);
  -ms-transform: translateY(50%) rotate(45deg);
  transform: translateY(50%) rotate(45deg);
  box-shadow: 1px 1px 0 0 rgba(0,0,0,0.2);
}

.c0:hover .c4 {
  display: block;
}

.c2 {
  margin: 0;
  padding: 1rem;
  color: #000000;
  background-color: #ffffff;
  border: 1px solid #d8d8d8;
  box-shadow: 0 3px 6px rgba(0,0,0,0.16),0 3px 6px rgba(0,0,0,0.23);
  display: none;
  position: absolute;
  z-index: 99;
  width: 100px;
  height: initial;
  left: 0;
  bottom: 0;
  -webkit-transform: translateY(-56px);
  -ms-transform: translateY(-56px);
  transform: translateY(-56px);
}

.c0:hover .c2 {
  display: block;
}

.c1 {
  box-sizing: border-box;
  position: relative;
  display: inline-block;
}

<div>
  <div
    className="c0 c1"
  >
    <h1>
      Trigger
    </h1>
    <div
      className="c2"
    >
      <p>
        Content
      </p>
      <div
        className="c3"
      />
    </div>
  </div>
</div>
`;
